File: 30daysOFcode_Go/day 12.go
package main

import "fmt"

/* Hackerrank does not support Go for problem 12 - inheritance*/

type human interface {
	greet()
}

// If a type implements the human interface, then print their details
func printPerson(h human) {
	switch h.(type) {
	case person:
		fmt.Printf("Name: %v, %v\n", h.(person).firstName, h.(person).lastName)
		fmt.Printf("ID: %v\n", h.(person).idNumber)
	case student:
		fmt.Printf("Name: %v, %v\n", h.(student).firstName, h.(student).lastName)
		fmt.Printf("ID: %v\n", h.(student).idNumber)
	}
}

type person struct {
	firstName, lastName string
	idNumber            int
}

// Method for type person that implements the human interface
func (p person) greet() {
	fmt.Printf("Hi, my name is %s\n", p.firstName)
}

type student struct {
	firstName, lastName string
	idNumber            int
	scores              []int
}

// Method for type student that implements the human interface
func (s student) greet() {
	fmt.Printf("Hi, my name is %s\n", s.firstName)
}

// Method to calculate a students average score
func (s student) calculate() int {
	total := 0
	n := len(s.scores)

	for _, val := range s.scores {
		total += val
	}

	return total / n
}

func main() {
	john := person{"John", "Smith", 13}
	printPerson(john)

	sally := student{
		firstName: "Sally",
		lastName:  "Smith",
		idNumber:  19,
		scores:    []int{100, 80},
	}

	printPerson(sally)
	avg := sally.calculate()
	fmt.Println("Sally's average score is:", avg)
}

File: 30daysOFcode_Python/day 12.py
class Person:
	def __init__(self, firstName, lastName, idNumber):
		self.firstName = firstName
		self.lastName = lastName
		self.idNumber = idNumber
	def printPerson(self):
		print("Name:", self.lastName + ",", self.firstName)
		print("ID:", self.idNumber)

class Student(Person):
    def __init__(self,firstName,lastName,idNum,scores):
        self.firstName=firstName
        self.lastName=lastName
        self.idNumber=idNum
        self.scores=scores

    def calculate(self):
        a=round(sum(self.scores)/len(self.scores))
        if a >= 90:
            return 'O'
        elif 90>a>=80:
            return 'E'
        elif 80>a>=70:
            return 'A'
        elif 70>a>=55:
            return 'P'
        elif 55>a>=40:
            return 'D'
        else:
            return 'T'

line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input()) # not needed for Python
scores = list( map(int, input().split()) )
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())